File: src/csharp/BelgianBeers/BelgianBeers/Tests/D03_PatternMatching.cs
using System;
using System.Linq;
using BelgianBeers.Models;
using Xunit;
using Xunit.Abstractions;
using static BelgianBeers.Tests.Utilities.TestData;

namespace BelgianBeers.Tests
{
    [Collection("Demo 3 - Pattern Matching")]
    public class D03_PatternMatching
    {
        private readonly ITestOutputHelper _outputHelper;

        public D03_PatternMatching(ITestOutputHelper outputHelper)
        {
            _outputHelper = outputHelper;
        }
        
        [Fact]
        public void PatternMatching()
        {
            // Get beers that are from brewery "Westmalle"
            var westmalleBeers = BeerFlow
                .Where(beer => string.Equals(beer.Brewery?.Name, "Brouwerij der Trappisten van Westmalle", StringComparison.OrdinalIgnoreCase))
                .ToList();
            //https://github.com/dotnet/csharplang/blob/master/proposals/nullable-reference-types.md
            // Pattern matching (on a property, not on type):
            foreach (var westmalleBeer in westmalleBeers.Select(PatchBeer))
            {
                switch (westmalleBeer)
                {
                    case DubbelBeer dubbelBeer:
                        // It is a dubbel
                        _outputHelper.WriteLine(dubbelBeer.Name);
                        break;
                    
                    case TripelBeer tripelBeer when westmalleBeer.Name.IndexOf("tripel", StringComparison.OrdinalIgnoreCase) >= 0:
                        // It is a tripel
                        _outputHelper.WriteLine(tripelBeer.Name);
                        break;
                }
            }
            
            Assert.True(westmalleBeers.Any());
        }

        private Beer PatchBeer(Beer beer)
        {
            if (beer.Name.IndexOf("dubbel", StringComparison.OrdinalIgnoreCase) >= 0)
            {
                return new DubbelBeer(beer.Name, beer.Brewery, beer.Rating, beer.Votes);
            }
            
            if (beer.Name.IndexOf("tripel", StringComparison.OrdinalIgnoreCase) >= 0)
            {
                return new TripelBeer(beer.Name, beer.Brewery, beer.Rating, beer.Votes);
            }

            return beer;
        }
    }
}

File: src/kotlin/BelgianBeers/common/src/test/kotlin/D03_PatternMatching.kt
import org.jetbrains.kotlincsharpdemo.Beer
import org.jetbrains.kotlincsharpdemo.TestData
import kotlin.test.Test

sealed class BeerWithTaste(val beer: Beer)

class JustBeer(beer: Beer) : BeerWithTaste(beer) {
  override fun toString() = "JustBeer($beer)"
}

class DubbelBeer(beer: Beer) : BeerWithTaste(beer) {
  val dubbelName get() = beer.Name
  override fun toString() = "BubbelBeer($beer)"
}

class TrippelBeer(beer: Beer) : BeerWithTaste(beer) {
  override fun toString() = "TrippelBeer($beer)"
}

val TrippelBeer.tripperName get() = beer.Name

fun tasteBeer(beer: Beer): BeerWithTaste {
  return when {
    beer.Name.contains("dubbel", ignoreCase = true) -> DubbelBeer(beer)
    beer.Name.contains("tripel", ignoreCase = true) -> TrippelBeer(beer)
    else -> JustBeer(beer)
  }
}

infix fun String.looksLike(s: String) = this.contains(s, ignoreCase = true)


class D03_PatternMatching {
  @Test
  fun patternMatching() {

    val beersWithTaste = TestData.beerFlow.map(::tasteBeer).toList()

    beersWithTaste.filterIsInstance<TrippelBeer>().forEach {
      println("Tripper: $it")
    }

    beersWithTaste.filterIsInstance<DubbelBeer>().forEach {
      println("Dubbel: $it")
    }

    val westmalleBeers = beersWithTaste.filter { it.beer.Name looksLike "Brouwerij der Trappisten van Westmalle" }

    // Pattern matching (on a property, not on type):
    for (westmalleBeer in westmalleBeers) {
      when {
        westmalleBeer is DubbelBeer ->
          println(westmalleBeer.dubbelName)

        westmalleBeer is TrippelBeer && westmalleBeer.beer.Name.contains("tripel", ignoreCase = true) ->
          println(westmalleBeer.tripperName)

        else -> {
          //No beer today!
        }
      }
    }
  }
}
